server/views.py

```diff
@@ -71,113 +71,76 @@ def index(request):
         schedule_filter = "schedule__" + now_day + "__contains"
-        availability_qs = qs.filter(**{schedule_filter: 2})
-
-
-    # check option 3 before midnight
+        schedule_id = 2
     elif is_time_between(time(23, 00), time(00, 00)):
         schedule_filter = "schedule__" + now_day + "__contains"
-        availability_qs = qs.filter(**{schedule_filter: 3})
-
-    # check option 3 after midnight
-    elif is_time_between(time(00, 00), time(7, 00)):
+        schedule_id = 3
+    else:
         schedule_filter = "schedule__" + yesterday_day + "__contains"
-        availability_qs = qs.filter(**{schedule_filter: 3})
-
-    availability_now_id = []
-    if availability_qs != []:
-        for volunteer in availability_qs:
-            availability_now_id.append(volunteer.id)
-
-    if len(availability) == 0:
-        availability_qs = Volunteer.objects.all().all()
-
-    # union matchings from both categoties
-    match_qs = area_qs.union(language_qs)
+        schedule_id = 3
 
-    # if there were no matches display all and there are people available
-    if len(match_qs) == 0 and (not something_mark):
-        match_qs = Volunteer.objects.all()
+    if len(availability) != 0 and '' not in availability:
+        filter_options[schedule_filter] = schedule_id
 
-    availability_qs = (availability_qs)
-    match_qs = match_qs.intersection(availability_qs)
+    match_qs = Volunteer.objects.filter(q_option, **filter_options).annotate(appears_count=Count('helprequest'))
+    availability_now_id = match_qs.filter(**{schedule_filter: schedule_id}).values_list('id', flat=True)
 
     # ----- orders -----
     if 'field' in request.GET:
         field = request.GET.get('field')
-        field = "-" + field
         match_qs = match_qs.order_by(field)
 
-    # ----- check for each volunteer how much times he apper
-    appers_list = []
-    valid_certificates = []
-    for volunteer in match_qs:
-        appers_list.append(HelpRequest.objects.filter(helping_volunteer=volunteer).count())
-        volunteer.get_active_certificates()
-        valid_certificate = volunteer.get_active_certificates().first()
-        if valid_certificate is not None:
-            valid_certificates.append(valid_certificate.id)
-        else:
-            valid_certificates.append(-1)
+    # ----- build final data ----
+    paged_data = []
 
-    # make match qs to tuple
-    final_data = []
-    for i in range(0, len(match_qs)):
-        final_data.append((match_qs[i], appers_list[i], valid_certificates[i]))
+    paginator = Paginator(match_qs, RESULTS_IN_PAGE)
 
-    paginator = Paginator(final_data, RESULTS_IN_PAGE)
+    paged_data = paginator.page(page)
 
-    final_data = paginator.page(page)
+    final_data = []
+    for volunteer in paged_data:
+        valid_certificate = volunteer.get_active_certificates().first()
+        final_data.append((volunteer, valid_certificate.id if valid_certificate is not None else -1))
```

Comment:
slightly more elegant way of doing this:
```suggestion
        valid_certificate_id = volunteer.get_active_certificates().first().id or -1
        final_data.append((volunteer, valid_certificate_id))
```